week-07-rails-continued/day-02/module-01/readme.md

```diff
@@ -1 +1,259 @@
+
+  ```erb
+  <!-- app/views/users/new.html.erb -->
+
+  <%= form_for :user, url: '/users' do |f| %>
+    <%= f.email_field :email, placeholder: "Email", autofocus: true %>
+    <%= f.password_field :password, placeholder: "Password" %>
+    <%= f.submit "Sign Up" %>
+  <% end %>
+
+  ```
+
+  **Stretch Challenge:** Can you do this in a modal in the `site#index`? Would you need your `users#new` route or action any more?
+
+14. In the `users#create` action, create a user, then log them in by creating a new session. After the user is successfully created and the session is set, redirect to the user show page.
+
+  ```ruby
+  #
+  # app/controllers/users_controller.rb
+  #
+  class UsersController < ApplicationController
+
+    ...
+
+    def create
+      user = User.create(user_params)
+      session[:user_id] = user.id
+
+      # redirect_to "/users/#{user.id}"
+      # the line above can be refactored to use rails route helpers:
+      redirect_to user_path(user)
+    end
+
+    ...
+
+  end
+  ```
+
+  **Stretch Challenge:** Create a condition that checks if the user was saved correctly.
+
+  **Hint:** first build the user in memory with `User.new(user_params)`, then check `if user.save` proceed as normal `else` redirect to `"/users/new"` again. Run `rails g scaffold user email password` in a SEPARATE SAMPLE PROJECT for an example.
+
+15. The last step is to display attributes for the currently logged in user on the user show page. To do this, update the `users#show` method in the controller find a user in the database using the current `session[:user_id]`.
+
+  ```ruby
+  #
+  # app/controllers/users_controller.rb
+  #
+  class UsersController < ApplicationController
+
+    ...
+
+    def show
+      @user = User.find(session[:user_id])
+    end
+
+    ...
+
+  end
+  ```
+
+16. Now, we want to display data for the currently logged-in user in the view:
+
+  ```erb
+  <!-- app/views/users/show.html.erb -->
+
+  Welcome, <%= @user.email %>!
+  ```
+
+1. Lastly, add a `current_user` method at the application level and then try to use it in your splash page.
+
+  ```ruby
+    #  app/controllers/application_controller.rb
+
+    class ApplicationController < ActionController::Base
+      protect_from_forgery with: :exception
+
+      def current_user
+        @current_user ||= session[:user_id] && User.find_by_id(session[:user_id])
+      end
+
+      helper_method :current_user  #makes the current_user method available to views
+    end
+  ```
+
+  ```erb
+  <!-- splash.html.erb -->
+  
+  Welcome <%= current_user %>
+  ```
+
+## Stretch Challenges
+
+1. Create a logout action and allow users to logout. What do you have to do to the `session[:user_id]`? Where do you `redirect_to` after they log out.
```

Comment:
I would move this into the "Challenges" section above, instead of having it as a stretch, since ideally they'll have both Sign Up, Log In, and Log Out done by the end of class.